README.md
# BigInt support for Acorn

[![NPM version](https://img.shields.io/npm/v/acorn-bigint.svg)](https://www.npmjs.org/package/acorn-bigint)

This is a plugin for [Acorn](http://marijnhaverbeke.nl/acorn/) - a tiny, fast JavaScript parser, written completely in JavaScript.

It implements support for arbitrary precision integers as defined in the stage 4 proposal [BigInt: Arbitrary precision integers in JavaScript](https://github.com/tc39/proposal-bigint). The emitted AST follows [ESTree](https://github.com/estree/estree/blob/master/es2020.md#bigintliteral).

## Usage

This module provides a plugin that can be used to extend the Acorn `Parser` class:

```javascript
const {Parser} = require('acorn');
const bigInt = require('acorn-bigint');
Parser.extend(bigInt).parse('100n');
```

## License

This plugin is released under an [MIT License](./LICENSE).


CHANGELOG.md
## 1.0.0 (2020-06-11)

* Remove `n` from bigint property (according to estree)

## 0.4.0 (2019-04-04)

* Make compatible with acorn-numeric-separator

## 0.3.1 (2018-10-06)

* Fix creation of BigInt values everywhere (Thanks, Gus Caplan!)

## 0.3.0 (2018-09-14)

* Update to new acorn 6 interface
* Actually support creating BigInt values in AST in Chrome
* Change license to MIT

## 0.2.0 (2017-12-20)

* Emit BigInt values in AST if supported by runtime engine

## 0.1.0 (2017-12-19)

Initial release


test/test.js
"use strict"

const assert = require("assert")
const acorn = require("acorn")
const Parser = acorn.Parser.extend(require(".."))

function test(text, expectedResult, additionalOptions) {
  it(text, function () {
    const result = Parser.parse(text, Object.assign({ ecmaVersion: 9 }, additionalOptions))
    assert.deepStrictEqual(result.body[0], expectedResult)
  })
}
function testFail(text, expectedError, additionalOptions) {
  it(text, function () {
    let failed = false
    try {
      Parser.parse(text, Object.assign({ ecmaVersion: 9 }, additionalOptions))
    } catch (e) {
      assert.strictEqual(e.message, expectedError)
      failed = true
    }
    assert(failed)
  })
}

const newNode = (start, props) => Object.assign(new acorn.Node({options: {}}, start), props)
const newBigIntLiteral = (start, stringValue) => newNode(start, {
  type: "Literal",
  end: start + stringValue.length + 1,
  value: typeof BigInt !== "undefined" ? BigInt(stringValue) : null,
  raw: `${stringValue}n`,
  bigint: stringValue
})

describe("acorn-bigint", function () {
  const digits = [
    {d: "0", ast: start => newBigIntLiteral(start, "0")},
    {d: "2", ast: start => newBigIntLiteral(start, "2")},
    {d: "0x2", ast: start => newBigIntLiteral(start, "0x2")},
    {d: "0o2", ast: start => newBigIntLiteral(start, "0o2")},
    {d: "0b10", ast: start => newBigIntLiteral(start, "0b10")},
    {d: "-0xbf2ed51ff75d380fd3be813ec6185780", ast: start => newNode(start, {
      type: "UnaryExpression",
      end: start + 36,
      operator: "-",
      prefix: true,
      argument: newBigIntLiteral(start + 1, "0xbf2ed51ff75d380fd3be813ec6185780")
    })},
    {d: "02", error: start => `Identifier directly after number (1:${start + 2})`},
    {d: "2e2", error: start => `Identifier directly after number (1:${start + 3})`},
    {d: "2.4", error: start => `Identifier directly after number (1:${start + 3})`},
    {d: ".4", error: start => `Identifier directly after number (1:${start + 2})`},
  ]
  const statements = [
    {s: "let i = %s", ast: content => newNode(0, {
      type: "VariableDeclaration",
      end: content.end,
      kind: "let",
      declarations: [newNode(4, {
        type: "VariableDeclarator",
        end: content.end,
        id: newNode(4, {
          type: "Identifier",
          end: 5,
          name: "i"
        }),
        init: content
      })]
    })},

    {s: "i = %s", ast: content => newNode(0, {
      type: "ExpressionStatement",
      end: content.end,
      expression: newNode(0, {
        type: "AssignmentExpression",
        end: content.end,
        operator: "=",
        left: newNode(0, {
          type: "Identifier",
          end: 1,
          name: "i"
        }),
        right: content
      })
    })},

    {s: "((i = %s) => {})", ast: content => newNode(0, {
      type: "ExpressionStatement",
      end: content.end + 8,
      expression: newNode(1, {
        type: "ArrowFunctionExpression",
        end: content.end + 7,
        id: null,
        generator: false,
        expression: false,
        async: false,
        params: [
          newNode(2, {
            type: "AssignmentPattern",
            end: content.end,
            left: newNode(2, {
              type: "Identifier",
              end: 3,
              name: "i"
            }),
            right: content
          })
        ],
        body: newNode(content.end + 5, {
          type: "BlockStatement",
          end: content.end + 7,
          body: []
        })
      })
    })},

    {s: "for (let i = 0n; i < %s;++i) {}", ast: content => newNode(0, {
      type: "ForStatement",
      end: content.end + 8,
      init: newNode(5, {
        type: "VariableDeclaration",
        end: 15,
        declarations: [
          newNode(9, {
            type: "VariableDeclarator",
            end: 15,
            id: newNode(9, {
              type: "Identifier",
              end: 10,
              name: "i"
            }),
            init: newBigIntLiteral(13, "0")
          })
        ],
        kind: "let"
      }),
      test: newNode(17, {
        type: "BinaryExpression",
        end: content.end,
        left: newNode(17, {
          type: "Identifier",
          start: 17,
          end: 18,
          name: "i"
        }),
        operator: "<",
        right: content
      }),
      update: newNode(content.end + 1, {
        type: "UpdateExpression",
        end: content.end + 4,
        operator: "++",
        prefix: true,
        argument: newNode(content.end + 3, {
          type: "Identifier",
          end: content.end + 4,
          name: "i"
        })
      }),
      body: newNode(content.end + 6, {
        type: "BlockStatement",
        end: content.end + 8,
        body: []
      })
    })},

    {s: "i + %s", ast: content => newNode(0, {
      type: "ExpressionStatement",
      end: content.end,
      expression: newNode(0, {
        type: "BinaryExpression",
        end: content.end,
        left: newNode(0, {
          type: "Identifier",
          start: 0,
          end: 1,
          name: "i"
        }),
        operator: "+",
        right: content
      })
    })}
  ]
  statements.forEach(statement => {
    const start = statement.s.indexOf("%s")
    digits.forEach(d => {
      (d.error ? testFail : test)(
        statement.s.replace("%s", `${d.d}n`),
        d.error ? d.error(start) : statement.ast(d.ast(start))
      )
    })
  })
})


inject.js
"use strict"

module.exports = function (acorn) {
  const tt = acorn.tokTypes
  const isIdentifierStart = acorn.isIdentifierStart

  acorn.plugins.bigInt = function (instance) {
    instance.extend("parseLiteral", function (superF) {
      return function(value) {
        const node = superF.call(this, value)
        if (node.raw.charCodeAt(node.raw.length - 1) == 110) node.bigint = node.raw
        return node
      }
    })

    instance.extend("readRadixNumber", function (_superF) {
      return function(radix) {
        let start = this.pos
        this.pos += 2 // 0x
        let val = this.readInt(radix)
        if (val === null) this.raise(this.start + 2, `Expected number in radix ${radix}`)
        if (this.input.charCodeAt(this.pos) == 110) {
          let str = this.input.slice(start, this.pos)
          val = typeof BigInt !== "undefined" && BigInt.parseInt ? BigInt.parseInt(str) : null
          ++this.pos
        } else if (isIdentifierStart(this.fullCharCodeAtPos())) this.raise(this.pos, "Identifier directly after number")
        return this.finishToken(tt.num, val)
      }
    })

    instance.extend("readNumber", function (superF) {
      return function(startsWithDot) {
        let start = this.pos

        // Not an int
        if (startsWithDot) return superF.call(this, startsWithDot)

        // Legacy octal
        if (this.input.charCodeAt(start) === 48 && this.input.charCodeAt(start + 1) !== 110) {
          return superF.call(this, startsWithDot)
        }

        if (this.readInt(10) === null) this.raise(start, "Invalid number")

        // Not a BigInt, reset and parse again
        if (this.input.charCodeAt(this.pos) != 110) {
          this.pos = start
          return superF.call(this, startsWithDot)
        }

        let str = this.input.slice(start, this.pos)
        let val = typeof BigInt !== "undefined" && BigInt.parseInt ? BigInt.parseInt(str) : null
        ++this.pos
        return this.finishToken(tt.num, val)
      }
    })
  }
  return acorn
}


run_test262.js
"use strict"

const path = require("path")
const run = require("test262-parser-runner")
const acorn = require("acorn")
const bigInt = require(".")
const Parser = acorn.Parser.extend(bigInt)

const unsupportedFeatures = ["object-rest", "object-spread", "regexp-named-groups",
  "async-iteration", "class-fields", "class-fields-public",
  "computed-property-names", // Only used for class fields
  "regexp-unicode-property-escapes",
  "regexp-lookbehind", "regexp-dotall", "optional-catch-binding"]

run(
  (content, options) => Parser.parse(content, {sourceType: options.sourceType, ecmaVersion: 9}),
  {
    testsDirectory: path.dirname(require.resolve("test262/package.json")),
    skip: test => (!test.attrs.features || !test.attrs.features.includes("BigInt") || unsupportedFeatures.some(f => test.attrs.features.includes(f))),
  }
)


index.js
"use strict"

const acorn = require("acorn")
const tt = acorn.tokTypes
const isIdentifierStart = acorn.isIdentifierStart

module.exports = function(Parser) {
  return class extends Parser {
    parseLiteral(value) {
      const node = super.parseLiteral(value)
      if (node.raw.charCodeAt(node.raw.length - 1) == 110) node.bigint = this.getNumberInput(node.start, node.end - 1)
      return node
    }

    readRadixNumber(radix) {
      let start = this.pos
      this.pos += 2 // 0x
      let val = this.readInt(radix)
      if (val === null) this.raise(this.start + 2, `Expected number in radix ${radix}`)
      if (this.input.charCodeAt(this.pos) == 110) {
        let str = this.getNumberInput(start, this.pos)
        val = typeof BigInt !== "undefined" ? BigInt(str) : null
        ++this.pos
      } else if (isIdentifierStart(this.fullCharCodeAtPos())) this.raise(this.pos, "Identifier directly after number")
      return this.finishToken(tt.num, val)
    }

    readNumber(startsWithDot) {
      let start = this.pos

      // Not an int
      if (startsWithDot) return super.readNumber(startsWithDot)

      // Legacy octal
      if (this.input.charCodeAt(start) === 48 && this.input.charCodeAt(start + 1) !== 110) {
        return super.readNumber(startsWithDot)
      }

      if (this.readInt(10) === null) this.raise(start, "Invalid number")

      // Not a BigInt, reset and parse again
      if (this.input.charCodeAt(this.pos) != 110) {
        this.pos = start
        return super.readNumber(startsWithDot)
      }

      let str = this.getNumberInput(start, this.pos)
      let val = typeof BigInt !== "undefined" ? BigInt(str) : null
      ++this.pos
      return this.finishToken(tt.num, val)
    }

    // This is basically a hook for acorn-numeric-separator
    getNumberInput(start, end) {
      if (super.getNumberInput) return super.getNumberInput(start, end)
      return this.input.slice(start, end)
    }
  }
}
